Date parsing failure

## Description

We have been using 4.0.2, trying to upgrade to 5.0.1 we hit this exception when querying /info. This is happening with both docker 1.7 and 1.11. Then downgrading to 4.0.8 as thought this was a change in the 5.x branch only we hit the same issue.

Suspecting this could be related to this commit?

https://github.com/spotify/docker-client/commit/49341c70949894be8dc27690e6a7d8ff5ed40cda

But this has been there for some time, so wondering if this hasn't been an issue for anyone else.

## How to reproduce

Call DockerClient::info

## What do you expect

[Describe what do you expect to happen]

## What happened instead

Exception:

Caused by: com.fasterxml.jackson.databind.exc.InvalidFormatException: Can not construct instance of java.util.Date from String value '2016-06-03T06:57:17.4782869-04:0
0': not a valid representation (error: Failed to parse Date value '2016-06-03T06:57:17.4782869-04:00': Can not parse date "2016-06-03T06:57:17.4782869-04:00": not com
patible with any of standard forms ("yyyy-MM-dd'T'HH:mm:ss.SSSZ", "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", "EEE, dd MMM yyyy HH:mm:ss zzz", "yyyy-MM-dd"))
 at [Source: org.glassfish.jersey.message.internal.ReaderInterceptorExecutor$UnCloseableInputStream@6f7ec518; line: 1, column: 756] (through reference chain: com.spot
ify.docker.client.messages.Info["SystemTime"])
        at com.fasterxml.jackson.databind.exc.InvalidFormatException.from(InvalidFormatException.java:55)
        at com.fasterxml.jackson.databind.DeserializationContext.weirdStringException(DeserializationContext.java:907)
        at com.fasterxml.jackson.databind.deser.std.StdDeserializer._parseDate(StdDeserializer.java:787)

## Software:

- docker version: 
1.11.1-cs1

as well as:

1.7.1

- docker-client version: [Add docker-client version here]

--- a/src/main/java/com/spotify/docker/client/DockerDateFormat.java
+++ b/src/main/java/com/spotify/docker/client/DockerDateFormat.java
@@ -38,14 +38,20 @@ public class DockerDateFormat extends StdDateFormat {
 
   private static final long serialVersionUID = 249048552876483658L;
 
+  // either a literal Z or an offset like -04:00 or +1200 with an optional colon separator
+  private static final String TIMEZONE_PATTERN = "(Z|[+-]\\d{2}:?\\d{2})";
+
   @Override
   public Date parse(String source) throws ParseException {
     // If the date has nanosecond precision (e.g. 2014-10-17T21:22:56.949763914Z), remove the last
     // digits so we can create a Date object, which only support milliseconds.
     // Docker doesn't always return nine digits for the fractional seconds part,
     // so we need to be more flexible when trimming to milliseconds.
-    if (source.matches(".+\\.\\d{4,9}Z$")) {
-      source = source.replaceAll("(\\.\\d{3})\\d{1,6}Z$", "$1Z");
+    // Also allow for flexible timezone formats, either 'Z' for zulu/UTC or hour offsets.
+    // StdDateFormat has logic for handling these other timezones, but only if the time portion of
+    // the string matches hh:mm:ss or hh:mm:ss.SSS
+    if (source.matches(".+\\.\\d{4,9}" + TIMEZONE_PATTERN + "$")) {
+      source = source.replaceAll("(\\.\\d{3})\\d{1,6}" + TIMEZONE_PATTERN + "$", "$1$2");
     }
 
     return super.parse(source);

--- a/src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java
+++ b/src/test/java/com/spotify/docker/client/DefaultDockerClientTest.java
@@ -1609,19 +1609,8 @@ public class DefaultDockerClientTest {
   @Test
   public void testDockerDateFormat() throws Exception {
     // This is the created date for busybox converted from nanoseconds to milliseconds
-
     final Date expected = new StdDateFormat().parse("2015-09-18T17:44:28.145Z");
-    final DockerDateFormat dateFormat = new DockerDateFormat();
-    // Verify DockerDateFormat handles millisecond precision correctly
-    final Date milli = dateFormat.parse("2015-09-18T17:44:28.145Z");
-    assertThat(milli, equalTo(expected));
-    // Verify DockerDateFormat converts nanosecond precision down to millisecond precision
-    final Date nano = dateFormat.parse("2015-09-18T17:44:28.145855389Z");
-    assertThat(nano, equalTo(expected));
-    // Verify DockerDateFormat converts nanosecond precision with less than nine digits
-    // down to millisecond precision
-    final Date nanoSevenDigits = dateFormat.parse("2015-09-18T17:44:28.1458553Z");
-    assertThat(nanoSevenDigits, equalTo(expected));
+
     // Verify the formatter works when used with the client
     sut.pull(BUSYBOX_BUILDROOT_2013_08_1);
     final ImageInfo imageInfo = sut.inspectImage(BUSYBOX_BUILDROOT_2013_08_1);

--- a/src/test/java/com/spotify/docker/client/DockerDateFormatTest.java
+++ b/src/test/java/com/spotify/docker/client/DockerDateFormatTest.java
@@ -19,6 +19,8 @@ package com.spotify.docker.client;
 
 import com.fasterxml.jackson.databind.util.StdDateFormat;
 
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
 import org.junit.Before;
 import org.junit.Test;
 
@@ -61,5 +63,12 @@ public class DockerDateFormatTest {
   @Test
   public void testHandlesNanosecondWithLessThanNineDigits() throws Exception {
     assertThat(dockerDateFormat.parse("2015-09-18T17:44:28.1458553Z"), equalTo(expected));
+  }
+
+  @Test
+  public void otherTimeZones() throws Exception {
+    final Date expected =
+        new DateTime(2016, 6, 3, 6, 57, 17, 478, DateTimeZone.forOffsetHours(-4)).toDate();
+    assertThat(dockerDateFormat.parse("2016-06-03T06:57:17.4782869-04:00"), equalTo(expected));
   }
 }

--- a/src/test/java/com/spotify/docker/client/DockerDateFormatTest.java
+++ b/src/test/java/com/spotify/docker/client/DockerDateFormatTest.java
@@ -1,0 +1,65 @@
+/*
+ * Copyright (c) 2016
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.spotify.docker.client;
+
+import com.fasterxml.jackson.databind.util.StdDateFormat;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.Date;
+
+import static org.hamcrest.Matchers.equalTo;
+import static org.junit.Assert.assertThat;
+
+public class DockerDateFormatTest {
+
+  private final DockerDateFormat dockerDateFormat = new DockerDateFormat();
+  private final String millisecondDateString = "2015-09-18T17:44:28.145Z";
+  private Date expected;
+
+  @Before
+  public void setUp() throws Exception {
+    expected = new StdDateFormat().parse(millisecondDateString);
+  }
+
+  /**
+   * Verify DockerDateFormat handles millisecond precision correctly
+   */
+  @Test
+  public void testHandlesMillisecondPrecision() throws Exception {
+    assertThat(dockerDateFormat.parse(millisecondDateString), equalTo(expected));
+  }
+
+  /**
+   * Verify DockerDateFormat converts nanosecond precision down to millisecond precision
+   */
+  @Test
+  public void testHandlesNanosecondPrecision() throws Exception {
+    assertThat(dockerDateFormat.parse("2015-09-18T17:44:28.145855389Z"), equalTo(expected));
+  }
+
+  /**
+   * Verify DockerDateFormat converts nanosecond precision with less than nine digits
+   * down to millisecond precision
+   */
+  @Test
+  public void testHandlesNanosecondWithLessThanNineDigits() throws Exception {
+    assertThat(dockerDateFormat.parse("2015-09-18T17:44:28.1458553Z"), equalTo(expected));
+  }
+}

